x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrations.java

```diff
@@ -95,7 +95,17 @@ private void updateRolesByQuery(
                 .execute(UpdateByQueryAction.INSTANCE, updateByQueryRequest, ActionListener.wrap(bulkByScrollResponse -> {
                     logger.info("Migrated [" + bulkByScrollResponse.getTotal() + "] roles");
                     listener.onResponse(null);
-                }, listener::onFailure));
+                }, (exception) -> {
+                    if (exception instanceof IllegalArgumentException
+                        && exception.getMessage() != null
+                        && exception.getMessage().contains("script_lang not supported [painless]")) {
```

Comment:
Why do we need this? The update by query with the script should be executed only when there are really roles that need migrating, no? Are there such cases, also with the painless plugin not installed?

In any case, this doesn't look like the right approach to me. It's adding main code for test purposes, and it's not clear what the state of the migration is in actuality when this is encountered.
I think we should either properly handle the case where the migration is not applicable (because the painless module is missing) or simply install the painless plugin where necessary in the ITs encountering it.

Reply:
>Why do we need this? The update by query with the script should be executed only when there are really roles that need migrating, no? Are there such cases, also with the painless plugin not installed?

Because these tests create roles before/in parallel with the migration and therefore the search for roles to migrate finds them. The only indicator we have that a role needs to be migrated or not is the `metadata` field, but since it's not indexed we can't run `exists` on it (so we can't improve the search query as far as I can see). 

>In any case, this doesn't look like the right approach to me. It's adding main code for test purposes, and it's not clear what the state of the migration is in actuality when this is encountered.

Yes, this only happens for these tests, but I agree, it's not ideal at all. 

>or simply install the painless plugin where necessary in the ITs encountering it.

We can't per my understanding. This happens because of the same issue you and Ryan are discussing here: https://elastic.slack.com/archives/C8UUBNASY/p1702482296432399?thread_ts=1701962741.577089&cid=C8UUBNASY

The options I've been working on so far: 

1. Make sure the security migrations do not run for these tests (write to cluster state in `@Before`, not great).
2. Remove the dependency on painless (a custom script service, feels like that's overkill).
3. Catch script_lang not supported [painless] and ignore it in the migration code (current implementation). 

Another option might be to try to figure out if an index manager state change is the result of the index being created, but couldn't get that reliable before.

Reply:
> Because these tests create roles before/in parallel with the migration and therefore the search for roles to migrate finds them. The only indicator we have that a role needs to be migrated or not is the metadata field, but since it's not indexed we can't run exists on it (so we can't improve the search query as far as I can see).

I see, thanks for explaining.
Can we wait for the migration in a @Before method of the test suite class? Can we add a setting to disable migrations altogether (for test suites that we know migrations will be noops anyways)?

Reply:
Yes, I can investigate that. The problem is that the `.security` index is created by the role creation that's part of the test. The creation of the index in turn triggers the migration.

Reply:
I ended up adding code to detect if the index is new or not, if it's new don't apply any migrations. This should cover all the test cases we currently have.